json-schema-provider/src/main/java/io/confluent/kafka/schemaregistry/json/JsonSchema.java

```diff
@@ -66,87 +69,116 @@
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Invalid JSON " + schemaString, e);
+    }
+  }
+
   public JsonSchema(Schema schemaObj) {
     this(schemaObj, null);
   }
 
   public JsonSchema(Schema schemaObj, Integer version) {
-    this.schemaObj = schemaObj;
-    this.version = version;
-    this.references = Collections.emptyList();
-    this.resolvedReferences = Collections.emptyMap();
+    try {
+      this.jsonNode = objectMapper.readTree(schemaObj.toString());
+      this.schemaObj = schemaObj;
+      this.version = version;
+      this.references = Collections.emptyList();
+      this.resolvedReferences = Collections.emptyMap();
+    } catch (IOException e) {
+      throw new IllegalArgumentException("Invalid JSON " + schemaObj.toString(), e);
+    }
   }
 
   private JsonSchema(
+      JsonNode jsonNode,
       Schema schemaObj,
       Integer version,
       List<SchemaReference> references,
       Map<String, String> resolvedReferences,
       String canonicalString
   ) {
+    this.jsonNode = jsonNode;
     this.schemaObj = schemaObj;
     this.version = version;
     this.references = references;
     this.resolvedReferences = resolvedReferences;
     this.canonicalString = canonicalString;
   }
 
-  public static JsonSchema copy(JsonSchema schema) {
+  public JsonSchema copy() {
     return new JsonSchema(
-        schema.schemaObj,
-        schema.version,
-        schema.references,
-        schema.resolvedReferences,
-        schema.canonicalString
+        this.jsonNode,
+        this.schemaObj,
+        this.version,
+        this.references,
+        this.resolvedReferences,
+        this.canonicalString
     );
   }
 
-  public static JsonSchema copy(JsonSchema schema, Integer version) {
+  public JsonSchema copy(Integer version) {
     return new JsonSchema(
-        schema.schemaObj,
+        this.jsonNode,
+        this.schemaObj,
         version,
-        schema.references,
-        schema.resolvedReferences,
-        schema.canonicalString
+        this.references,
+        this.resolvedReferences,
+        this.canonicalString
     );
   }
 
   public Schema rawSchema() {
+    if (schemaObj == null) {
+      try {
+        SchemaLoader.SchemaLoaderBuilder builder = SchemaLoader.builder();
+        for (Map.Entry<String, String> dep : resolvedReferences.entrySet()) {
+          builder.registerSchemaByURI(new URI(dep.getKey()), new JSONObject(dep.getValue()));
+        }
+        JSONObject jsonObject = objectMapper.treeToValue((jsonNode), JSONObject.class);
+        builder.schemaJson(jsonObject);
+        SchemaLoader loader = builder.build();
+        schemaObj = loader.load().build();
+      } catch (URISyntaxException e) {
+        throw new IllegalArgumentException("Invalid dependency name", e);
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Invalid JSON", e);
+      }
+    }
```

Comment:
Is this ever accessed concurrently ? If so, we should protect the 'schemaObj' field when accessing it. This was likely not a problem before since the field was set in the constructor.

Reply:
This class is not meant to be accessed concurrently.

Reply:
Got it !